python/taichi/graph/_graph.py

```diff
@@ -97,77 +99,227 @@ def run(self, args):
+
+    if tag == ArgKind.SCALAR:
+        if "element_shape" in kwargs:
+            warnings.warn(
+                "The element_shape argument for scalar will be deprecated in v1.5.0. You can remove them safely.",
+                DeprecationWarning)
+            del kwargs["element_shape"]
 
     if tag == ArgKind.NDARRAY:
-        # Ndarray with matrix data type
-        if isinstance(dtype, MatrixType):
-            return _ti_core.Arg(tag, name, dtype.dtype, ndim,
-                                dtype.get_shape())
-        # Ndarray with scalar data type
-        if len(element_shape) > 0:
+        if "element_shape" not in kwargs:
+            if "dtype" in kwargs:
+                dtype = kwargs["dtype"]
+                if isinstance(dtype, MatrixType):
+                    kwargs["dtype"] = dtype.dtype
+                    kwargs["element_shape"] = dtype.get_shape()
+                else:
+                    kwargs["element_shape"] = ()
+        else:
             warnings.warn(
                 "The element_shape argument for ndarray will be deprecated in v1.5.0, use vector or matrix data type instead.",
                 DeprecationWarning)
-        return _ti_core.Arg(tag, name, dtype, ndim, element_shape)
+            if "dtype" not in kwargs:
+                dtype = kwargs["dtype"]
+                if isinstance(dtype, MatrixType):
+                    raise TaichiRuntimeError(
+                        "Please do not specify element_shape when dtype is a matrix type."
+                    )
 
-    if tag == ArgKind.MATRIX:
-        if not isinstance(dtype, MatrixType):
-            raise TaichiRuntimeError(
-                f'Tag {tag} must specify matrix data type, but got {dtype}.')
-        if len(element_shape) > 0:
-            raise TaichiRuntimeError(
-                f'Element shape for MatrixType argument "{name}" is not supported.'
-            )
-        mat_type = dtype
-        arg_list = []
-        i = 0
-        for _ in range(mat_type.n):
-            arg_sublist = []
-            for _ in range(mat_type.m):
-                arg_sublist.append(
-                    _ti_core.Arg(tag, f'{name}_mat_arg_{i}', dtype.dtype, ndim,
-                                 element_shape))
-                i += 1
-            arg_list.append(arg_sublist)
-        return arg_list
-
-    if tag == ArgKind.TEXTURE or tag == ArgKind.RWTEXTURE:
-        if channel_format is None or len(shape) == 0 or num_channels is None:
+    if tag == ArgKind.RWTEXTURE or tag == ArgKind.TEXTURE:
+        if "dtype" in kwargs:
+            warnings.warn(
+                "The dtype argument for texture will be deprecated in v1.5.0, use format instead.",
+                DeprecationWarning)
+            del kwargs["dtype"]
+
+        if "shape" in kwargs:
+            warnings.warn(
+                "The shape argument for texture will be deprecated in v1.5.0, use ndim instead. (Note that you no longer need the exact texture size.)",
```

Comment:
OOC: can various textures with the same shape share the same taichi kernel?

Reply:
yes, we have struct for over rw textures